rust-src/concordium_base/src/curve_arithmetic/mod.rs

```diff
@@ -18,6 +19,85 @@ use thiserror::Error;
 pub enum CurveDecodingError {
     #[error("Not a point on the curve.")]
     NotOnCurve,
+    #[error("{0} is not a field element.")]
+    NotInField(String),
+}
+
+/// This trait represents an element of a field.
+pub trait Field: Sized + Eq + Copy + Clone + Send + Sync + fmt::Debug + fmt::Display {
+    /// Returns an element chosen uniformly at random using a user-provided RNG.
+    fn random<R: RngCore + ?std::marker::Sized>(rng: &mut R) -> Self;
+
+    /// Returns the zero element of the field, the additive identity.
+    fn zero() -> Self;
+
+    /// Returns the one element of the field, the multiplicative identity.
+    fn one() -> Self;
+
+    /// Returns true iff this element is zero.
+    fn is_zero(&self) -> bool;
+
+    /// Squares this element.
+    fn square(&mut self);
+
+    /// Doubles this element.
+    fn double(&mut self);
+
+    /// Negates this element.
+    fn negate(&mut self);
+
+    /// Adds another element to this element.
+    fn add_assign(&mut self, other: &Self);
+
+    /// Subtracts another element from this element.
+    fn sub_assign(&mut self, other: &Self);
+
+    /// Multiplies another element by this element.
+    fn mul_assign(&mut self, other: &Self);
+
+    /// Computes the multiplicative inverse of this element, if nonzero.
+    fn inverse(&self) -> Option<Self>;
+
+    /// Exponentiates this element by a number represented with `u64` limbs,
+    /// least significant digit first.
+    fn pow<S: AsRef<[u64]>>(&self, exp: S) -> Self {
+        let mut res = Self::one();
+
+        let mut found_one = false;
+
+        for i in ff::BitIterator::new(exp) {
+            if found_one {
+                res.square();
+            } else {
+                found_one = i;
+            }
+
+            if i {
+                res.mul_assign(self);
+            }
+        }
+
+        res
+    }
+}
+
+/// This is an extension of the `Field` trait that adds some constants decribing
+/// the element size and operations for conveting to/from bib integer
+/// representation (an array of `u64` limbs.)
+pub trait PrimeField: Field {
+    /// How many bits are needed to represent an element of this field.
+    const NUM_BITS: u32;
+
+    /// How many bits of information can be reliably stored in the field
+    /// element.
+    const CAPACITY: u32;
+
+    /// Get a big integer representation with least significant digit first.
+    fn into_repr(self) -> Vec<u64>;
```

Comment:
We can in principle make this tighter, we could define 
```
const NUM_LIMBS
```
as part of the trait and then have

```
into_repr(self) -> [u64; NUM_LIMBS]
```

But maybe we can do this in a followup PR.

Reply:
Yes, that's true. However, we cannot make `NUM_LIMBS` an associated constant and use it in the type. This is an experimental feature currently. We could abstract the representation type, but this requires looking at all usage points to see how exactly we use the representation. We might need to add some constraints on this type. So, let's keep it as it is and looke at it at some later point.